### PR TITLE
Feature/db2 temp tables

### PR DESCRIPTION
**Pull request by detule**
This is an attempt at finishing off https://github.com/r-dbi/odbc/pull/461

* Rebased on top of main
* Rather than customizing both `dbListTables` and `dbExistsTable`, opted to specialize `odbcConnectionTables`.  I think this resulted in a much lighter touch.

More detail on temporary tables in DB2 [here](https://www.ibm.com/docs/en/db2-for-zos/12?topic=statements-declare-global-temporary-table).

```
> conn@info
$dbname
[1] "SAMPLE"

$dbms.name
[1] "DB2/LINUXX8664"

$db.version
[1] "11.05.0800"
...
$drivername
[1] "libdb2.a"

$odbcdriver.version
[1] "03.51"
....

DBI::dbExecute(conn, sqlCreateTable(conn, "testtbl", df, temporary = TRUE))
[1] -1
DBI::dbExecute(conn, sqlCreateTable(conn, "testtbl2", df, temporary = TRUE))
[1] -1
dbExistsTable(conn, name = "testtbl")
[1] TRUE
dbExistsTable(conn, name = "testtbl", schema_name = "SESSION")
[1] TRUE
dbExistsTable(conn, name = "testtbl2")
[1] TRUE
dbExistsTable(conn, name = "testtblnothere")
[1] FALSE
dbListTables(conn, schema_name="SESSION")
[1] "testtbl2" "testtbl"
```

### CODE REVIEW

File: R/driver-db2.R [hunk 1]
```diff
@@ -29,3 +30,44 @@ setMethod("sqlCreateTable", "DB2/AIX64",
...
+           exact = FALSE) {
+
+    res <- callNextMethod()
+    dfTempTables <- NULL
+    if ( (is.null(schema_name) || schema_name == "SESSION") &&
+         (is.null(catalog_name) || tolower(catalog_name) == conn@info$username) ) {
+      # Also look for temp tables in this case
+      qTable <- getSelector("TABNAME", name, exact)
+      qSchema <- getSelector("TABSCHEMA", schema_name, exact)
+      query <- paste0(
```
Comment (hadley) [inline R/driver-db2.R:55]: I think you should include a link to the docs in a comment here.

File: R/driver-db2.R [hunk 2]
```diff
@@ -29,3 +30,44 @@ setMethod("sqlCreateTable", "DB2/AIX64",
...
+  function(conn,
+           name,
+           catalog_name = NULL,
+           schema_name = NULL,
+           table_type = NULL,
+           exact = FALSE) {
+
+    res <- callNextMethod()
+    dfTempTables <- NULL
+    if ( (is.null(schema_name) || schema_name == "SESSION") &&
```
Comment (hadley) [inline R/driver-db2.R:50]: I almost wonder if it's worth adding a `odbConnectionTempTables()` generic that we could call from `odbcConnectionTables()`. Not for this PR, but might be interesting to consider in the future.

File: R/driver-db2.R [hunk 3]
```diff
@@ -29,3 +30,44 @@ setMethod("sqlCreateTable", "DB2/AIX64",
...
+      query <- paste0(
+        "SELECT
+            INSTANTIATOR AS \"table_catalog\",
+            TABSCHEMA AS \"table_schema\",
+            TABNAME AS \"table_name\",
+            'TABLE' AS \"table_type\",
+            NULL AS \"table_remarks\"
+         FROM SYSIBMADM.ADMINTEMPTABLES ",
+        "WHERE 1 = 1 ", qTable, qSchema)
+      tryCatch({
```
Comment (hadley) [inline R/driver-db2.R:64]: Why do we need a `tryCatch` here? Could we scope it more narrowly to a subclass of error?

**Review by hadley — APPROVED** [review]

File: NEWS.md [hunk 4]
```diff
@@ -1,6 +1,9 @@
 # odbc (development version)
 
+* DB2: Better support for temp tables in `dbListTables`, and `dbExistsTable`.
+
 * oracle: Fix writing to DATE and TIMESTAMP(n) targets using `batch_size` > 1.
+  (#810)
```
Comment (simonpcouch) [inline NEWS.md:6]:
```suggestion
* oracle: Fix writing to DATE and TIMESTAMP(n) targets using `batch_size` > 1 (#810).
```

[Tidy style](https://style.tidyverse.org/news.html#general-style) nit :)

File: NEWS.md [hunk 5]
```diff
@@ -1,6 +1,9 @@
 # odbc (development version)
 
+* DB2: Better support for temp tables in `dbListTables`, and `dbExistsTable`.
```
Comment (simonpcouch) [inline NEWS.md:3]:
```suggestion
* DB2: Better support for temp tables in `dbListTables`, and `dbExistsTable` (#823).
```

**Review by simonpcouch — APPROVED** [review]
Stellar🏄‍♀️

File: R/driver-db2.R (hunk 1 above)
Comment (detule) [reply R/driver-db2.R:59]: Done / thanks.

File: R/driver-db2.R (hunk 2 above)
Comment (detule) [reply R/driver-db2.R:50]: Yep  - we've needed to do this a few times now.  Probably a good idea.

File: R/driver-db2.R (hunk 3 above)
Comment (detule) [reply R/driver-db2.R:68]: I am no `DB2` expert, but I was a bit concerned that not every user might have access to tables/views prefixed with `ADMIN`.

Not sure if the approach I took in the follow up is what you had in mind - subclass `rethrow_database_error::cli_abort` - let me know if not.

File: NEWS.md (hunk 5 above)
Comment (detule) [reply NEWS.md]: Thanks Simon / Done

File: NEWS.md (hunk 4 above)
Comment (detule) [reply NEWS.md]: Done / thanks.

**Pull request merged**